src/include/execution/util/stage_timer.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <vector>
+
+#include "execution/util/macros.h"
+#include "execution/util/timer.h"
+
+namespace terrier::util {
+
+/**
+ * Stage timer
+ * @tparam ResolutionRatio resolution of the timer
+ */
+template <typename ResolutionRatio = std::milli>
+class StageTimer {
+ public:
+  /**
+   * Information about a stage.
+   */
+  class Stage {
+   public:
+    /**
+     * Create a new stage with name @em name.
+     * @param name The name of the stage.
+     */
+    explicit Stage(const char *name) noexcept : name_(name), time_(0) {}
+
+    /**
+     * Return the name of this stage.
+     */
+    const char *name() const noexcept { return name_; }
+
+    /**
+     * Return the time this stage took in the configured resolution ratio.
+     */
+    double time() const noexcept { return time_; }
+
+   private:
+    friend class StageTimer<ResolutionRatio>;
+
+    void set_time(const double time) { time_ = time; }
+
+   private:
+    // The name
+    const char *const name_;
+    // The time in this stage
+    double time_;
+  };
+
+  /**
+   * Construct.
+   */
+  StageTimer() { stages_.reserve(8); }
```

Comment:
It isn't hardcoded, it's just reserving to optimize the common case, i.e., fewer than eight stages.

Reply:
It is hard-coding the reserved size.  I'm suggesting that rather than assuming the common case is 8, you take the reservation size as a parameter to the constructor.  At the very least, we should document what the "common case" is as this will likely get co-opted elsewhere in the system once we start trying to instrument more code for metrics collection.